datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -369,11 +366,8 @@ impl CaseExpr {
         // evaluate when expression
         let when_value = self.when_then_expr[0].0.evaluate(batch)?;
         let when_value = when_value.into_array(batch.num_rows())?;
-        let when_value = as_boolean_array(&when_value).map_err(|e| {
-            DataFusionError::Context(
```

Comment:
Another potential way to to keep the context would be something like

```rust
.map_err(|e| e.context("WHEN expression did not return a BooleanArray"))
```

I think what you have here is good though as it makes it clear this is an internal / non expected error